service/history/nDCTransactionMgr.go

```diff
@@ -232,83 +244,119 @@ func (r *nDCTransactionMgrImpl) backfillWorkflow(
 			}
-		} else if mode == persistence.UpdateWorkflowModeUpdateCurrent {
-			// case 2
-			transactionPolicy = transactionPolicyActive
+			return persistence.UpdateWorkflowModeUpdateCurrent, transactionPolicyActive, nil
+		}
+
+		// need to reset target workflow (which is also the current workflow)
+		// to accept events to be reapplied
+		baseMutableState := targetWorkflow.getMutableState()
+		domainID := baseMutableState.GetExecutionInfo().DomainID
+		workflowID := baseMutableState.GetExecutionInfo().WorkflowID
+		baseRunID := baseMutableState.GetExecutionInfo().RunID
+		resetRunID := uuid.New()
+		baseRebuildLastEventID := baseMutableState.GetPreviousStartedEventID()
+
+		// TODO when https://github.com/uber/cadence/issues/2420 is finished
+		//  remove this block
+		if baseRebuildLastEventID == common.EmptyEventID {
 			r.logger.Warn("cannot reapply event to a finished workflow",
-				tag.WorkflowDomainID(targetWorkflowEvents.DomainID),
-				tag.WorkflowID(targetWorkflowEvents.WorkflowID),
+				tag.WorkflowDomainID(domainID),
+				tag.WorkflowID(workflowID),
 			)
 			r.metricsClient.IncCounter(metrics.HistoryReapplyEventsScope, metrics.EventReapplySkippedCount)
-			// TODO when https://github.com/uber/cadence/issues/2420 is finished
-			//  reset to workflow finish event and reapply to new resetted workflow by using
-			//  transactionPolicyActive, ignore this case for now
-		} else {
-			// target workflow is active but not being pointed by current record
-			// do not need to handle events reapplication here
+			return persistence.UpdateWorkflowModeBypassCurrent, transactionPolicyPassive, nil
 		}
+
+		baseVersionHistories := baseMutableState.GetVersionHistories()
+		baseCurrentVersionHistory, err := baseVersionHistories.GetCurrentVersionHistory()
+		if err != nil {
+			return 0, transactionPolicyActive, err
+		}
+		baseRebuildLastEventVersion, err := baseCurrentVersionHistory.GetEventVersion(baseRebuildLastEventID)
+		if err != nil {
+			return 0, transactionPolicyActive, err
+		}
+		baseCurrentBranchToken := baseCurrentVersionHistory.GetBranchToken()
+		baseNextEventID := baseMutableState.GetNextEventID()
+
+		if err = r.workflowResetter.resetWorkflow(
+			ctx,
+			domainID,
+			workflowID,
+			baseRunID,
+			baseCurrentBranchToken,
+			baseRebuildLastEventID,
+			baseRebuildLastEventVersion,
+			baseNextEventID,
+			resetRunID,
+			uuid.New(),
+			targetWorkflow,
+			eventsReapplicationResetWorkflowReason,
+			targetWorkflowEvents.Events,
+		); err != nil {
+			return 0, transactionPolicyActive, err
+		}
+		// after the reset of target workflow (current workflow) with additional events to be reapplied
+		// target workflow is no longer the current workflow
+		return persistence.UpdateWorkflowModeBypassCurrent, transactionPolicyPassive, nil
 	}
 
-	return targetWorkflow.getContext().updateWorkflowExecutionWithNew(
-		now,
-		mode,
-		nil,
-		nil,
-		transactionPolicy,
-		nil,
-	)
+	if err := targetWorkflow.getContext().reapplyEvents(
```

Comment:
reading through the code, its not obvious to me as to what we do here i.e. at this point, either targetWorkflow is not current workflow or this is not the active cluster. What do we now ? Can you add a comment to help the reader ?

Reply:
sure